### PR TITLE
Hi-res, downscale, and video recording wrappers

### PR DESCRIPTION
**Pull request by shwang**
Video example (seizure warning?): https://youtu.be/XSjXbIbf-pU

For hi-res image and/or downsampling examples (and also image diffs) see https://docs.google.com/document/d/1tXk4nr6sYMMozUaauFREg2IQCue5nKWemPG6KTh8yt0/edit#heading=h.sb0u39tuqhky

### CODE REVIEW

File: minerl/herobraine/wrappers/downscale_wrapper.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+from minerl.herobraine.hero.spaces import Box, Dict
+
+
+class DownscaleWrapper(ObservationWrapper):
+    def __init__(self, env, downscale_res=(64, 64)):
+        super().__init__(env)
+        self.downscale_res = downscale_res
+
+        obs_space = env.observation_space
+
```
Comment (shwang) [inline minerl/herobraine/wrappers/downscale_wrapper.py]: Add check that we are even multiple of the env pov dimensions

File: minerl/herobraine/wrappers/high_res_wrapper.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        self.resolution = resolution
+        super().__init__(env_to_wrap)
+
+    def _update_name(self, name: str) -> str:
+        task, ver = name.split('-')
+        return task + "HighRes-" + ver
+
+    def create_observables(self):
+        observables = super().create_observables()
+        new_pov = handlers.POVObservation(self.resolution)
```
Comment (shwang) [inline minerl/herobraine/wrappers/high_res_wrapper.py]: assert that we change exactly one observable handler

File: minerl/herobraine/wrappers/video_recording_wrapper.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        elif self.outdir:
+            if not os.path.exists(self.outdir):
+                os.makedirs(self.outdir)
+            target_file = os.path.join(self.outdir, str(self.current_index) + ".mp4")
+            self.current_index += 1
+            write_mp4(self.images, self.outfile)
+
+
+def write_mp4(images, target_file):
+    fourcc = cv2.VideoWriter_fourcc(*'mp4v')
```
Comment (shwang) [inline minerl/herobraine/wrappers/video_recording_wrapper.py:48]: (seems equivalent to `VideoWriter_fourcc('m', 'p', '4', 'v')`, which is weird. Confirm this)

File: minerl/herobraine/wrappers/video_recording_wrapper.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    # Need to change the order of the input dimensions for CV2
+    out = cv2.VideoWriter(target_file, fourcc, 20.0, (images[0].shape[1], images[0].shape[0]))
```
Comment (shwang) [inline minerl/herobraine/wrappers/video_recording_wrapper.py]: Is there a `with` block equivalent that calls `release()`?

Comment (shwang) [reply minerl/herobraine/wrappers/video_recording_wrapper.py]: Could also do `try-finally`.

File: minerl/herobraine/wrappers/video_recording_wrapper.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+    for image in images:
+        out.write(np.flip(image, axis=-1))  # cv2 uses BGR
+    out.release()
+    cv2.destroyAllWindows()
```
Comment (shwang) [inline minerl/herobraine/wrappers/video_recording_wrapper.py]: (why do we need to destroy all windows?)

File: minerl/herobraine/wrappers/video_recording_wrapper.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        return obs, rew, done, info
+
+    def reset(self):
+        if self.images:
+            self.save_video()
+            self.images = []
+        return super().reset()
+
+    def save_video(self):
+        if self.outfile:
```
Comment (shwang) [inline minerl/herobraine/wrappers/video_recording_wrapper.py]: Maybe refactor into `self.get_next_outfile`

File: minerl/herobraine/wrappers/video_recording_wrapper.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+        """
+        super().__init__(env)
+        assert outfile or outdir, "Must include somewhere to save the video."
+        assert not (outfile and outdir), "Can only choose one way to save videos."
+        self.outfile = outfile
+        self.outdir = outdir
+        # Used for directory saving.
+        self.current_index = 0
+
```
Comment (shwang) [inline minerl/herobraine/wrappers/video_recording_wrapper.py:51]: Fail fast by checking whether 'pov' exists in the observation space dict

File: minerl/herobraine/wrappers/downscale_wrapper.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+        obs_dict = {key: obs_space[key] for key in obs_space}
+        # This is specifically tailored for the MineRL observations...
+        obs_dict['pov'] = Box(0, 255, downscale_res + (3,),
+                              dtype=np.uint8)
+        self.observation_space = Dict(obs_dict)
+
+    def observation(self, obs):
```
Comment (shwang) [inline minerl/herobraine/wrappers/downscale_wrapper.py:21]: - [x] Figure out why we are getting the wrong dimensions (either at this wrapper layer or at the EnvSpec HighRes wrapper layer.

File: minerl/herobraine/env_specs/basalt_specs.py [hunk 9]
```diff
@@ -95,15 +94,28 @@ def _basalt_gym_entrypoint(
 
 class BasaltBaseEnvSpec(EnvSpec):
 
+    LOW_RES_SIZE = 64
+    # HIGH_RES_SIZE = 1024
```
Comment (shwang) [inline minerl/herobraine/env_specs/basalt_specs.py]: Change this back or settle on odd-number before final review

File: minerl/herobraine/wrappers/video_recording_wrapper.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+            (self.video_height, self.video_width),
+        )
+
+    def write_rgb_image(self, img: np.ndarray):
+        assert img.ndim == 3
+        self.out.write(np.flip(img, axis=-1))
+
+    def close(self):
+        if self.out is not None:
+            self.out.release()
```
Comment (brandonhoughton) [inline minerl/herobraine/wrappers/video_recording_wrapper.py:38]: Just checking - is release the correct close action here?

Comment (shwang) [reply minerl/herobraine/wrappers/video_recording_wrapper.py:38]: Yeah, `cv2` is a C++ bindings package, and the naming convention there is `release()`

**Review by brandonhoughton — APPROVED** [review]

**Comment by shwang** [comment]
Sweet this has been a long time coming 🎉🎉🎉

**Pull request merged**